documentation/typescript/adding-custom-login-form-with-spring-security.asciidoc

```diff
@@ -0,0 +1,176 @@
+
+.frontend/login-view.ts
+[source, typescript]
+----
+@customElement('login-view')
+export class LoginView extends LitElement implements AfterEnterObserver {
+
+  @property({type: Boolean})
+  private error = false;
+
+  @property({type: Boolean})
+  private open = true;
+
+  // the url to redirect to after a successful login
+  private returnUrl = '/';
+
+  private onSuccess: (result:LoginResult) => void;
+
+  constructor(){
+    super();
+    this.onSuccess = () => {
+      Router.go(this.returnUrl);
+    };
+  }
+
+  render() {
+    return html`
+      <vaadin-login-overlay
+        ?opened="${this.open}" 
+        .error=${this.error}
+        @login="${this.login}">    
+      </vaadin-login-overlay>
+    `;
+  }
+
+  async login(event: CustomEvent): Promise<LoginResult> {
+    this.error = false;
+    // use the Vaadin provided login helper method to obtain the LoginResult
+    const result = await login(event.detail.username, event.detail.password);
+    this.error = result.error;
+  
+    if (!result.error) {
+      this.onSuccess(result);
+    }
+
+    return result;
+  }
+
+  onAfterEnter(location: RouterLocation) {
+    this.returnUrl = location.redirectFrom || this.returnUrl;
+  }
+}
+----
+After the login view is defined, you should also define a route for the login view component, for example in the `index.ts` file. Note, the `path` for the login view component should be the same as what you defined in `SecurityConfig` for `http.formLogin().loginPage()`.
+.frontend/index.ts
+[source, typescript]
+----
+const routes = [
+  {
+    path: '/login',
+    component: 'login-view'
+  },
+  // more routes
+}
+----
+== Redirect the user to the login view
+To protect a view from unauthenticated users, that is, redirect unauthenticated users to the login view, you can use route action.
+
+.frontend/index.ts
+[source, typescript]
+----
+const isUserLoggedIn = function() {
+  return localStorage.getItem('loggedIn') === 'true';
+}
+
+const routes = [
+  ...
+  {
+    path: '/my-view',
+    action: async (_: Router.Context, commands: Router.Commands) => {
+      if (!await isUserLoggedIn()) {
+        return commands.redirect('/login');
+      }
+      return undefined;
+    },
+    component: 'my-view'
+  }
+  ...
+}
+----
+You can also add the route action to the parent layout, so that all the child views are protected.
+
+.frontend/index.ts
+[source, typescript]
+----
+const routes = [
+  ...
+  {
+    path: '/main-layout',
+    action: async (_: Router.Context, commands: Router.Commands) => {
+      if (!await isUserLoggedIn()) {
+        return commands.redirect('/login');
+      }
+      return undefined;
+    },
+    component: 'main-layout',
+    children: [
+      ...
+    ]
+  }
+  ...
+}
+----
+The `isUserLoggedIn()` method in the above code examples uses
+a `loggedIn` boolean stored in the https://vaadin.com/components/browse[localstorage] to check if the user is logged in. The boolean needs to be reset when logging out.
```

Comment:
Why is there a link to https://vaadin.com/components/browse here? Should it maybe link to MDN docs about localStorage?

Reply:
Also "localstorage" should be written as "localStorage".

Reply:
It would be good to include some explanation of why we're using a locally stored boolean for checking logged in status (so that it works offline too) and that this is not a security issue because it is only used to decide which views will be shown from the views that are already available in the client. And include some note that the content included directly in the TS views files (behind authentication) should not be considered secure (unless the whole view is loaded dynamically after authenticating to the server) and all secure data needs to be fetched via endpoints where the server makes sure the user is authenticated.

Or do we have this kind of explanation somewhere else already?

Reply:
The link was a bad copy past mistake, fixed.
Add a brief explanation why it was used. Note this is for V18, which doesn't have all the offline feature we are developing, so prefer not to expose much to it. For it to work better offline, would also need to use some timestamp, will update this once offline is to be merged to master. 
Regarding security, I think it's explained in this [page](https://vaadin.com/docs/v18/flow/typescript/configuring-security.html).

Reply:
@haijian-vaadin that page seems to be mainly about endpoint security. It doesn't tell anything about the security of content in the TS view files.

My main concern is that someone sees our example for adding authentication to a view and then has some information in the view (not fetched via endpoint, but hardcoded into the view) that is not supposed to be seen by non-authenticated users. So it should be explained somewhere that this way is not secure. Because the user could just set the loggedIn value in localStorage themselves to see all the views. But it will probably take some time to write a meaningful section about this and could be left outside the scope of this PR imo.

Reply:
I see, very good point, this could be some best practices article/tutorial with security. Let's do it in another article.